packages/gatsby/src/internal-plugins/functions/gatsby-node.ts

```diff
@@ -0,0 +1,184 @@
+import fs from "fs-extra"
+import glob from "globby"
+import path from "path"
+import webpack from "webpack"
+import multer from "multer"
+import * as express from "express"
+
+import { urlResolve } from "gatsby-core-utils"
+
+import { internalActions } from "../../redux/actions"
+
+import { ParentSpanPluginArgs, CreateDevServerArgs } from "gatsby"
+
+import TerserPlugin from "terser-webpack-plugin"
+
+export async function onPreBootstrap({
+  reporter,
+  store,
+}: ParentSpanPluginArgs): Promise<void> {
+  const activity = reporter.activityTimer(`Compiling Gatsby Functions`)
+  activity.start()
+
+  const {
+    program: { directory: siteDirectoryPath },
+    functions,
+  } = store.getState()
+
+  const functionsGlob = `**/*.{js,ts}`
+
+  const functionsDirectoryPath = path.join(siteDirectoryPath, `src/api`)
+
+  const functionsDirectory = path.resolve(
+    siteDirectoryPath,
+    functionsDirectoryPath as string
+  )
+
+  const files = await glob(functionsGlob, { cwd: functionsDirectory })
+
+  if (files?.length === 0) {
+    reporter.warn(
+      `No functions found in directory: ${path.relative(
+        siteDirectoryPath,
+        functionsDirectory
+      )}`
+    )
+  }
+
+  reporter.verbose(`Attaching functions to development server`)
+
+  const knownFunctions = new Map(
+    files.map(file => [
+      urlResolve(path.parse(file).dir, path.parse(file).name),
+      file,
+    ])
+  )
+
+  store.dispatch(internalActions.setFunctions(knownFunctions))
+
+  await fs.ensureDir(path.join(siteDirectoryPath, `.cache`, `functions`))
+
+  await fs.emptyDir(path.join(siteDirectoryPath, `.cache`, `functions`))
+
+  const gatsbyVarObject = Object.keys(process.env).reduce((acc, key) => {
+    if (key.match(/^GATSBY_/)) {
```

Comment:
IIRC if we don't need the match result `/^GATSBY_/.test(key)` will perform better

Reply:
fixed